### PR TITLE
Access code_typet::return_type() via code_typet's API

### PR DESCRIPTION
**Pull request by tautschnig**
Using the API properly hides an implementation detail.

- [x] Each commit message has a non-empty body, explaining why the change was made.
- [x] My contribution is formatted in line with CODING_STANDARD.md.
- [ ] Methods or procedures I have added are documented, following the guidelines provided in CODING_STANDARD.md.
- [ ] Regression or unit tests are included, or existing tests cover the modified code (in this case I have detailed which ones those are in the commit message).
- [ ] My commit message includes data points confirming performance improvements (if claimed).
- [x] My PR is restricted to a single feature or bugfix.
- [x] White-space or formatting changes outside the feature-related changed lines are in commits of their own.

### CODE REVIEW

File: src/cpp/cpp_declarator_converter.cpp [hunk 1]
```diff
@@ -326,11 +326,10 @@ void cpp_declarator_convertert::handle_initializer(
 {
   exprt &value=declarator.value();
 
-  // moves member initializers into 'value'
-  cpp_typecheck.move_member_initializers(
-    declarator.member_initializers(),
-    symbol.type,
-    value);
+  // moves member initializers into 'value' - only methods have these
+  if(!declarator.member_initializers().get_sub().empty())
```
Comment (smowton) [inline src/cpp/cpp_declarator_converter.cpp]: You add an initializer... only if there is already one?

**Review by smowton — COMMENTED** [review]
One query re: the apparent fix interleaved with the style improvement

File: src/cpp/cpp_declarator_converter.cpp (hunk 1 above)
Comment (tautschnig) [reply src/cpp/cpp_declarator_converter.cpp]: The method called will actually only do anything when this condition evaluates to true. Not checking this here would invoke the method for non-code-type things, which makes the `to_code_type` fail, even though the invoked method would just be a no-op.

Comment (tautschnig) [reply src/cpp/cpp_declarator_converter.cpp]: So really not a change in behaviour, just a necessary condition to make the subsequent `to_code_type` (which is now necessary) pass.

Comment (smowton) [reply src/cpp/cpp_declarator_converter.cpp]: That suggests you should check whether it has id() == ID_code instead?

**Review by allredj — COMMENTED** [review]
Passed Diffblue compatibility checks (cbmc commit: 4fe05f1652d4af83fc6547efb0e4dab11d5789c9).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/85549105

File: src/cpp/cpp_declarator_converter.cpp (hunk 1 above)
Comment (tautschnig) [reply src/cpp/cpp_declarator_converter.cpp]: Indeed. Done.

**Review by smowton — APPROVED** [review]

**Review by allredj — COMMENTED** [review]
Passed Diffblue compatibility checks (cbmc commit: 584db9de8efbbd04987bce66a89fc8337838f2bf).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/85572929

File: src/ansi-c/c_typecheck_expr.cpp [hunk 2]
```diff
@@ -1999,7 +1998,7 @@ void c_typecheck_baset::typecheck_side_effect_function_call(
         new_symbol.location=expr.source_location();
         new_symbol.type=code_typet();
         new_symbol.type.set(ID_C_incomplete, true);
-        new_symbol.type.add(ID_return_type)=return_type;
+        to_code_type(new_symbol.type).return_type() = return_type;
```
Comment (kroening) [inline src/ansi-c/c_typecheck_expr.cpp]: simply do
```
new_symbol.type=code_typet({ }, return_type);
```
two lines above

File: src/cpp/cpp_typecheck.cpp [hunk 3]
```diff
@@ -208,7 +208,7 @@ void cpp_typecheckt::static_and_dynamic_initialization()
   init_symbol.mode=ID_cpp;
   init_symbol.module=module;
   init_symbol.type=code_typet();
-  init_symbol.type.add(ID_return_type)=typet(ID_constructor);
+  to_code_type(init_symbol.type).return_type() = typet(ID_constructor);
   init_symbol.is_type=false;
```
Comment (kroening) [inline src/cpp/cpp_typecheck.cpp]: same as above

**Review by kroening — APPROVED** [review]

File: src/cpp/cpp_typecheck.cpp (hunk 3 above)
Comment (tautschnig) [reply src/cpp/cpp_typecheck.cpp]: Both done, thanks.

**Review by allredj — COMMENTED** [review]
Passed Diffblue compatibility checks (cbmc commit: 8d742c6529a04b3c3006ea694182d63c6f7d5e84).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/85633511

**Pull request merged**